### ISSUE TITLE
FriendlyName and ValidationError::DUPLICATED_ATTRIBUTE_NAME_FOUND

### ISSUE THREAD

**Issue by willshepherdson** [SAML-Toolkits/php-saml#372]
I'm currently trying to use php-saml to connect a Shibboleth IdP with my application. When troubleshooting the integration, a `ValidationError::DUPLICATED_ATTRIBUTE_NAME_FOUND` error was thrown for `FriendlyName`. Here is a snippet of the Attribute Statement:

```xml
<saml2:Attribute FriendlyName="uid" Name="urn:oid:0.9.2342.19200300.100.1.1" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
        <saml2:AttributeValue xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="xsd:string">username</saml2:AttributeValue>
</saml2:Attribute>
<saml2:Attribute FriendlyName="uid" Name="uid" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
        <saml2:AttributeValue xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="xsd:string">username</saml2:AttributeValue>
</saml2:Attribute>
```

I understand the [position](https://github.com/onelogin/python3-saml/issues/39#issuecomment-262021759) that it is uncommon for there to be multiple instances with the same Name attribute. However, the [SAML2 attribute spec](https://docs.oasis-open.org/security/saml/SpecDrafts-Post2.0/sstc-saml-attribute-x500-cd-01.html) seems to indicate that that comparison may not be appropriate for the `FriendlyName` attribute.

> **2.3.1 Attribute Name Comparison**
> Two `<Attribute>` elements refer to the same SAML attribute if and only if their Name XML attribute values are equal in the sense of [RFC3061]. The `FriendlyName` attribute plays no role in the comparison.

When the `getAttributesWithFriendlyName` method was added to `processResponse`  (https://github.com/onelogin/php-saml/pull/270), an abstraction was added to allow `getAttributesByFriendlyName` and `getAttributes` to reuse that base function. This means that `processResponse` will throws a ValidationError if a `FriendlyName` attribute is duplicated in the XML, which I don't believe is expected behavior. However, the XML does validate properly against the OneLogin XML Validator.

Would you consider revisiting this behavior? Or do you recommend continuing to enforce distinct attribute names for both `Name` and `FriendlyName`?

**Comment by pitbulk** [comment]
The issue is that getAttributes and getAttributesWithFriendlyName retrieve an indexed array where the Name or the FriendlyName, if we don't raise the DUPLICATED_ATTRIBUTE_NAME_FOUND, how are we gonna manage your use case, what should be retrieved by getAttributesWithFriendlyName?

**Comment by noel-fallstrom-digicert** [comment]
We have received this error when customers were submitting what should be an array of values with multiple instances of the same attribute. For example:
```<saml2:Attribute Name="role">
        <saml2:AttributeValue>admin</saml2:AttributeValue>
</saml2:Attribute>
<saml2:Attribute Name="role">
        <saml2:AttributeValue>user</saml2:AttributeValue>
</saml2:Attribute>```

What we did to solve this was to remove the error for duplicates and to merge the results into an array when a duplicate key was found:
```if (array_key_exists($attributeKeyName, $attributes)) {
	$attributeValues = array_merge($attributes[$attributeKeyName], $attributeValues);
}```

**Comment by dejury** [comment]
I think this one can be closed now, since release 3.5.0.
